File: lib/rayz/color.ex
defmodule Rayz.Color do
  @moduledoc """
  Documentation for Rayz.Color.
  """

  @doc """
      iex> Builder.color(1.1, 2.2, 3.3)
      %Rayz.Color{red: 1.1, green: 2.2, blue: 3.3}
  """

  defstruct(
    red:   0.0,
    green: 0.0,
    blue:  0.0
  )

  @type color :: %Rayz.Color{red: float(), green: float(), blue: float()}

  @spec multiply(color, float()) :: color
  def multiply(color, scalar) when is_float(scalar) do
    %Rayz.Color{
      red:   color.red   * scalar,
      green: color.green * scalar,
      blue:  color.blue  * scalar
    }
  end

  @spec multiply(color, color) :: color
  def multiply(color1, color2) when is_map(color2) do
    # hadamard or Schur product
    %Rayz.Color{
      red:   color1.red   * color2.red,
      green: color1.green * color2.green,
      blue:  color1.blue  * color2.blue
    }
  end

  @spec add(color, color) :: color
  def add(color1, color2) do
    %Rayz.Color{
      red:   color1.red   + color2.red,
      green: color1.green + color2.green,
      blue:  color1.blue  + color2.blue
    }
  end

  @spec subtract(color, color) :: color
  def subtract(color1, color2) do
    %Rayz.Color{
      red:   color1.red   - color2.red,
      green: color1.green - color2.green,
      blue:  color1.blue  - color2.blue
    }
  end

  @spec equal?(color, color) :: boolean()
  def equal?(color1, color2) do
    Util.equal?(color1.red,   color2.red)   &&
    Util.equal?(color1.green, color2.green) &&
    Util.equal?(color1.blue,  color2.blue)
  end
end

File: features/step_definitions/color.rb
Given('c ← color\({float}, {float}, {float})') do |float, float2, float3|
  @c = Rayz::Color.new(red: float, green: float2, blue: float3)
end

Then("c.red = {float}") do |float|
  assert_equal(@c.red, float)
end

Then("c.green = {float}") do |float|
  assert_equal(@c.green, float)
end

Then("c.blue = {float}") do |float|
  assert_equal(@c.blue, float)
end

Given('color1 ← color\({float}, {float}, {float})') do |float, float2, float3|
  @color1 = Rayz::Color.new(red: float, green: float2, blue: float3)
end

Given('color2 ← color\({float}, {float}, {float})') do |float, float2, float3|
  @color2 = Rayz::Color.new(red: float, green: float2, blue: float3)
end

Given('color3 ← color\({float}, {float}, {float})') do |float, float2, float3|
  @color3 = Rayz::Color.new(red: float, green: float2, blue: float3)
end

Given('c1 ← color\({float}, {float}, {float})') do |float, float2, float3|
  @c1 = Rayz::Color.new(red: float, green: float2, blue: float3)
end

Given('c2 ← color\({float}, {float}, {float})') do |float, float2, float3|
  @c2 = Rayz::Color.new(red: float, green: float2, blue: float3)
end

Then('c1 + c2 = color\({float}, {float}, {float})') do |float, float2, float3|
  expected = Rayz::Color.new(red: float, green: float2, blue: float3)
  assert_equal(@c1 + @c2, expected)
end

Then('c1 - c2 = color\({float}, {float}, {float})') do |float, float2, float3|
  expected = Rayz::Color.new(red: float, green: float2, blue: float3)
  assert_equal(@c1 - @c2, expected)
end

Then('c1 * c2 = color\({float}, {float}, {float})') do |float, float2, float3|
  expected = Rayz::Color.new(red: float, green: float2, blue: float3)
  assert_equal(@c1 * @c2, expected)
end

Then('c * {float} = color\({float}, {float}, {float})') do |float, float2, float3, float4|
  expected = Rayz::Color.new(red: float2, green: float3, blue: float4)
  assert_equal(@c * float, expected)
end
